lib/bundle/checker.rb

```diff
@@ -0,0 +1,130 @@
+# frozen_string_literal: true
+
+module Bundle
+  module Checker
+    class Base
+      PACKAGE_TYPE = :pkg
+      PACKAGE_TYPE_NAME = "Package"
+
+      NO_ACTION = [].freeze
+
+      def action_required_for(formula)
+        [formula]
+      end
+
+      def exit_early_check(packages)
+        work_to_be_done = packages.find do |pkg|
+          yield pkg
+        end
+        if work_to_be_done
+          action_required_for(work_to_be_done)
+        else
+          NO_ACTION
+        end
+      end
+
+      def full_check(packages)
+        packages.reject { |f| installed_and_up_to_date? f }
+                .map { |entry| "#{self.class::PACKAGE_TYPE_NAME} #{entry} needs to be installed or updated." }
+      end
+
+      def checkable_entries(all_entries)
+        all_entries.select { |e| e.type == self.class::PACKAGE_TYPE }
+      end
+
+      def format_checkable(entries)
+        checkable_entries(entries).map(&:name)
+      end
+
+      def installed_and_up_to_date?(_pkg)
+        raise NotImplementedError
+      end
+
+      def find_actionable(entries)
+        requested = format_checkable entries
+
+        if Bundle::Commands::Check.exit_on_first_error?
+          exit_early_check(requested) { |pkg| !installed_and_up_to_date?(pkg) }
+        else
+          full_check requested
+        end
+      end
+    end
+
+    module_function
+
+    CheckResult = Struct.new :work_to_be_done, :completed_checks, :errors, :unchecked_checks
+
+    CHECKS = {
+      taps_to_tap: "Taps",
+      casks_to_install: "Casks",
+      apps_to_install: "Apps",
+      formulae_to_install: "Formulae",
+    }.freeze
+
+    def check(exit_on_first_error)
+      @dsl ||= Bundle::Dsl.new(Bundle.brewfile)
+
+      check_method_names = CHECKS.keys
+
+      completed_checks = []
+      errors = []
+      enumerator = exit_on_first_error ? :any? : :map
+
+      work_to_be_done = check_method_names.send(enumerator) do |check_method|
+        check_errors = send(check_method)
+        completed_checks << check_method
+        any_errors = check_errors.any?
+        errors.concat(check_errors) if any_errors
+        any_errors
+      end
+
+      work_to_be_done = work_to_be_done.any? if work_to_be_done.class == Array
```

Comment:
Sniffing the class name here feels a bit weird.

Reply:
Yeah… I did that a few days ago. It got the job done and I feel weird about it. I'd like to revisit it at some point. It's there because `any?` returns boolean and `map` returns an array, and the consumers of that value expect a boolean.

Maybe I should be using `find` instead of `any?` for the enumerator when exiting early? That'd keep the operation exiting at the first it finds, right? If I've always got an array, then I can drop the `if wtbd.class == Array` check.

Thoughts?

Reply:
Ah, no, because `find` returns the first result, not an array. But, what if I put in some array flexory?

```patch
diff --git a/lib/bundle/checker.rb b/lib/bundle/checker.rb
index 344f3cd..c4fc30a 100644
--- a/lib/bundle/checker.rb
+++ b/lib/bundle/checker.rb
@@ -70,7 +70,7 @@ module Bundle

       completed_checks = []
       errors = []
-      enumerator = exit_on_first_error ? :any? : :map
+      enumerator = exit_on_first_error ? :find : :map

       work_to_be_done = check_method_names.send(enumerator) do |check_method|
         check_errors = send(check_method)
@@ -80,7 +80,7 @@ module Bundle
         any_errors
       end

-      work_to_be_done = work_to_be_done.any? if work_to_be_done.class == Array
+      work_to_be_done = [work_to_be_done].flatten.any?

       unchecked_checks = (check_method_names - completed_checks)

```

I could put that mechanism into another method, perhaps called `ensure_array` or something like that. 

This way is a little nicer to read but creating an array just to flatten it vs checking types ¯\\\_(ツ)_/¯

Reply:
`Array(work_to_be_done).flatten.any?` will ensure it's an array without creating a new one.

Reply:
✅